add sample for copying only used part of hive file.

NiKiZe opened the pull request:
Make it cleaner to analyze, give more compression, and possible smaller filesize.
Includes example of how checksum is writen to regf block.

NiKiZe commented:
If there is any other parts of hives that can be cleared from data it should be added/removed as well.

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+#    This file is part of python-registry.
+#    It tries to make a clean copy of hive file
+#    clean meaning without unused (not understood) data
+#
```
would you mind providing a bit more documentation here, particularly for users who may not be familiar with the registy format?

EricZimmerman replied:
what is 'not understood' data?

williballenthin replied:
yeah, same question here. i think its "slack" data.

williballenthin replied:
sorry, "unallocated", not "slack".

NiKiZe replied:
windows leaves a lot of junk in there, and there is no documentation on not understood data.
Maybe something like this?
`It tries to make a clean copy of hive file removing parts that is not used (windows leaves some junk) This makes it easier to analyze real data in hives.`

EricZimmerman replied:
id have to see an example. there isnt any junk in a hive as determined by the header's hive length that i have seen. perhaps you are referring to data node records or slack in free cell data?

NiKiZe replied:
I will still call byte 0x71 to 0x1fb junk, even if it is a leftover stack.
During my checks there has not been any real data there, even if it might look like it.
The hbins and cells might have slack, but if i create a new hive, or export from an hive (cmd reg save) i expect these parts to be clean, or be an exact copy of the original which is not the case, for lack of better words I would like to call it random bytes.

To understand the internal I wrote a disector NiKiZe@1b34f83787cbe39de0f8969eab6691565a999290 running that on hive files it bacame somewhat painfully obvious. Since my goal is to create write support (or maybe a better word is build support) I want to concentrate on the used parts and not the random parts.

EricZimmerman replied:
ah you are talking header specific. yea theres junk in there =)

NiKiZe replied:
what I define as junk is also cells and hbins that has never been used but still have non zero bytes in them. It is incorrect to call them junk if there is rests of real records in there, but that is not my experience so far.

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+
+from __future__ import print_function
+
+import sys
```
i like to order imports by:

  - in standard library
  - installable via pip
  - local to the current project

and within each section, by module, and then line length. i'm not always perfect at it, but its good for consistency.

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+        Arguments:
+        - `offset`: The relative offset from the start of the block.
+        - `data`: The data to be writen, can be multiple
+        """
```
thanks for the good documentation

NiKiZe replied:
The credit should probably go back to you since it's mostly a copy of the existing functions.
Now when I think about how much was copied - i will add you as copyright header as well.

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+        - `offset`: The relative offset from the start of the block.
+        - `data`: The data to be writen, can be multiple
+        """
+        print("write dword 0x%08x at offset 0x%08x" % (data[0], self._offset + offset) )
```
i'd recommend using the `logging` module over `print`. it seems to save time in the long run when LOC grows and you have to start redirecting things or importing other functions.

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+        print("write qword 0x%016x at offset 0x%08x" % (data[0], self._offset + offset) )
+        self._writer.seek(offset)
+        self._writer.write(struct.pack(str("<Q"), *data))
+
```
one space between methods, two between classes or functions

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+            self.pack_dword(offset + 0x8, hbinsize)
+            self.pack_qword(offset + 0xc, hbin.unpack_qword(0xc)) # Unknown (mostly 0x0)
+            self.pack_qword(offset + 0x14, hbin.unpack_qword(0x14)) # timestamp
+            self.pack_dword(offset + 0x1c, hbin.unpack_dword(0x1c)) # unknown (mostly 0x0)
```
good documentation

williballenthin commented on samples/copyclean.py:
```diff
@@ -0,0 +1,129 @@
+        #write checksum to file
+        self.pack_dword(0x1FC, xsum)
+        
+        for hbin in self.hbins():
```
so will this reproduce the same header generated by Windows?

NiKiZe replied:
It will not keep "junk" and will calculate a valid checksum, don't know how picky windows is, but it works fine loading the hive after cleanup etc.
Suggestion on needed documentation? how about replacing L80 with
`#Read header including name, skipping unused data and recalculate checksum.`
And L84: `#copy fields as dwords for simple checksum calculation, even if some parts are strings.`

EricZimmerman replied:
i used more or less the same format in my code and it generates a matching CRC as stored in the header (when the hive is clean of course). https://github.com/EricZimmerman/Registry/blob/master/Registry/Other/RegistryHeader.cs#L50

NiKiZe replied:
True, this assumes the orignal hive is clean to begin with.
Maybe it should throw en exception if the original file is not clean?

williballenthin commented:
Thanks for sending this PR. I added some comments (many of them minor) to suggest formatting consistency and clarifications of comments.  I think the most important thing is more robust documentation that describes what this does, and why. Help a potential users understand "why?" they would want this script.

From what I can tell, it looks like you're stripping out unallocated cells and keeping the allocated ones. Might be interesting to generate a hive with the inverse, so we can work on recovering data.

NiKiZe commented:
Yes only copying the parts that is declared used and trying to "zero" everything else.
Not quite sure I understand the "generate a hive with the inverse"

Will try to provide better documentation (examples) on what this actually does so there is no missunderstanding. But will probably need help with actually writing user-friendly explanations thereafter ;)

williballenthin commented:
By inverse I mean, remove all the valid stuff, and see what is left. We should see all the unallocated space, which is where we'd expect to find deleted data and whatnot. Also, "junk" :-)

EricZimmerman commented:
be careful with this approach. you will miss things if you drop unfree structures as there are sometimes  deleted nk and vk records in the "slack" space of lists and other structures. by all means explore it tho. it will be interesting to compare your common format export once you support deleted keys and values.

NiKiZe commented:
Added an branch wit a bunch of test files, most interesting "disector diff" is probably this:
https://github.com/NiKiZe/python-registry/blob/tests/testing/resources/BCDefi.vscpy.disector.diff
Comparing the original BCD file with the "cleaned up" version.

This is most likely not interesting for any forensic purposes, However it might be a good help in developing other tools? Does this make sense?

NiKiZe commented:
Please let me know if there is something that can be better and I will be happy to change.
Before merge I would prefer to squash it all to one commit, let me know if it's OK and I'll do that.

NiKiZe commented:
Found a possible bug with this one.
If you "minify" a hive and then run this again, or if you run this on a hive created with `reg save`
Then you get this exception
```Python
Traceback (most recent call last):
  File "samples/copyclean.py", line 130, in <module>
    copy.copy()
  File "samples/copyclean.py", line 105, in copy
    for cell in hbin.cells():
  File "/home/nixze/python-registry/samples/Registry/RegistryParse.py", line 1346, in cells
    c = c.next()
  File "/home/nixze/python-registry/samples/Registry/RegistryParse.py", line 344, in next
    raise RegistryStructureDoesNotExist("HBINCell does not exist at 0x%x" % (self._offset + self.size()))
Registry.RegistryParse.RegistryStructureDoesNotExist: Registry Structure Does Not Exist Exception: HBINCell does not exist at 0x7000
```
Since the only used part is `first_hbin = next(self.hbins())` in `def first_key(self):` this have never been tested before I guess?
Right now my brain is in The "unable to think clearly" state. Will look into it at an later time. But my guess is an "of by one" error in `def has_next(self):`
However this does not affect the result since it is at end of hive when it happens and everything is already written.

NiKiZe commented:
Wow I was tired... Doh!
Functions `def cells(self):` and `def records(self):` are almost identical but with the only difference that records is protected with a `try ... except RegistryStructureDoesNotExist: break` to handle this problem?
No part of the project use these functions?

Most relevant history I found is: https://github.com/williballenthin/python-registry/commit/4a0ae8795955a0fb36be71820bfc8ae987bf2fdb#diff-dafc73e13536f60eb7fb50889d37e31fR586

Any insight into the history or external usage would be great!

By adding a small check, this can be avoided. The functions would then return exactly the same.
```diff
@@ -1391,6 +1391,8 @@ class HBINBlock(RegistryBlock):

         while c.offset() < self._offset_next_hbin:
             yield c
+            if c.offset() + c.size() == self._offset_next_hbin:
+                break
             c = c.next()
```

williballenthin commented:
I don't anything uses HBINBlock.records(). In fact, I don't think it should even exist (the method doesn't return Record instances like you might suggest, but cell instances). So if we can, I'd rather kill it than fix the bug. We'll have to note this clearly since we're changing a potentially public API.

williballenthin commented:
thanks for making the changes earlier based on my feedback. hopefully you're not finding me too picky of a project maintainer.  if i haven't made it clear yet, i'm really excited that you've gotten involved with the project and are experimenting with it! if my recommendations are confusing/annoying/bad, please do let me know, and i'm happy to explain them in more depth (or back off).

NiKiZe commented:
Agreed, looking in to deprecation warnings...
```
def records(self):
    from warnings import warn
    warn("records is obsolete, use cells instead!")
    return self.cells()
```
Or does this cause problems since it uses yield?
Should it still be documented?

williballenthin commented:
yeah, this is good. lets add an entry in the CHANGELOG.txt file with a big "[DEPRECATED]" tag. I like the use of `warnings` --- I didn't know about that. Also, redirecting to `.cells` is a good idea.

NiKiZe commented:
Not at all, glad you are picky! I strive for perfection, but it is hard to fix issues I'm not aware of, so just glad they are pointed out.
You might have noticed I'm quite picky about commit history. And as I learn more Python (can't say I have read the PEPs yet), my pickyness will grow as well :D

Will try to craft another PR for the cells/records issue.

NiKiZe commented:
Rebased on #48 since that fix is needed for this to work (will close that one)
Squashed changes - sorry if it makes it hard to review (let me know what you prefer) but it might be easier to spot any issues this way(?)
And @williballenthin @EricZimmerman Thanks for the feedback!

williballenthin commented on Registry/RegistryParse.py:
```diff
@@ -284,6 +284,13 @@ def hive_name(self):
         """
         return self.unpack_string(0x30, 64).decode("utf-16le").rstrip("\x00")
 
+    def first_hbin_offset(self):
```
good idea
